framework/src/media/MediaRecorderImpl.cpp

```diff
@@ -136,11 +136,13 @@ void MediaRecorderImpl::prepareRecorder(recorder_result_t& ret)
 	if (mCurState != RECORDER_STATE_IDLE || mOutputDataSource == nullptr) {
 		meddbg("prepare Failed mCurState: %d mOutputDataSource : %s\n", (recorder_state_t)mCurState, \
 			(mOutputDataSource == nullptr ? "nullptr" : "not null"));
+		ret = (mOutputDataSource == nullptr) ? RECORDER_INVALID_DATASOURCE : RECORDER_INVALID_STATE;
 		return notifySync();
 	}
 
 	if (!mOutputDataSource->open()) {
 		meddbg("open() failed\n");
+		ret = RECORDER_FILE_OPEN_FAILED;
```

Comment:
recorder doesn't know current datasource using file or not so below should be considered.
1. each of datasource's operation must return correct error
 a. If file open failed because there's no file -> PLAYER_ERROR_NO_SUCH_FILE
 b. If file creation failed because there's no space -> RECORDER_ERROR_NO_SPACE
 c. Both of player, recorder if it failed during establish connection -> XXX_ERROR_CONNECTION_FAILED
 ...
2. To do this, we need an extra logic like a errno of oops in tinyalsa

Reply:
I do not think the media framework needs to tell me more about file open errors.

Reply:
For example, We need it when we try to play mp3 file in file system.

Reply:
ok. i think need the new define for datasource,
after merged, i will define error for datasource internal.